[Mixed][Improvement]:Support for serialization and deserialization thread-safe in lookup JOIN

### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/amoro/issues?q=is%3Aissue) and found no similar issues.


### What would you like to be improved?

Serialize and deserialize operations are not thread-safe in the `BinaryRowDataSerializerWrapper`. 

Using the [`reuseRow` and `reuseWriter`](https://github.com/NetEase/amoro/blob/ca2061edcb117b472eb9e0a3c84933c37118b79e/mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/lookup/BinaryRowDataSerializerWrapper.java#L42C1-L42C1) to serialize the RowData could speed the serialization process up, but it may lead to serializing incorrectly in multi-threads.

Related logs:
```java
Caused by: java.lang.IndexOutOfBoundsException: offset: 0, length: 64, size: 48
	at org.apache.flink.core.memory.DataOutputSerializer.write(DataOutputSerializer.java:134) ~[plugin_ne-flink-1.14.0-1.0.13_scala2.12_hive2.1.1_arctic0.5.1.4-release-3.9.16-1.4.11.0.1.jar:?]
	at org.apache.flink.table.runtime.typeutils.BinaryRowDataSerializer.serializeWithoutLength(BinaryRowDataSerializer.java:146) ~[flink-table_2.12-ne-flink-1.14.0-1.0.13.jar:ne-flink-1.14.0-1.0.13]
	at org.apache.flink.table.runtime.typeutils.BinaryRowDataSerializer.serialize(BinaryRowDataSerializer.java:88) ~[flink-table_2.12-ne-flink-1.14.0-1.0.13.jar:ne-flink-1.14.0-1.0.13]
	at com.netease.arctic.flink.lookup.BinaryRowDataSerializerWrapper.serialize(BinaryRowDataSerializerWrapper.java:61) ~[plugin_ne-flink-1.14.0-1.0.13_scala2.12_hive2.1.1_arctic0.5.1.4-release-3.9.16-1.4.11.0.1.jar:?]
	at com.netease.arctic.flink.lookup.RocksDBCacheState.serializeKey(RocksDBCacheState.java:144) ~[plugin_ne-flink-1.14.0-1.0.13_scala2.12_hive2.1.1_arctic0.5.1.4-release-3.9.16-1.4.11.0.1.jar:?]
	at com.netease.arctic.flink.lookup.RocksDBSetSpilledState.serializeKey(RocksDBSetSpilledState.java:78) ~[plugin_ne-flink-1.14.0-1.0.13_scala2.12_hive2.1.1_arctic0.5.1.4-release-3.9.16-1.4.11.0.1.jar:?]
	at com.netease.arctic.flink.lookup.RocksDBSetSpilledState.get(RocksDBSetSpilledState.java:114) ~[plugin_ne-flink-1.14.0-1.0.13_scala2.12_hive2.1.1_arctic0.5.1.4-release-3.9.16-1.4.11.0.1.jar:?]
	at com.netease.arctic.flink.lookup.SecondaryIndexTable.get(SecondaryIndexTable.java:85) ~[plugin_ne-flink-1.14.0-1.0.13_scala2.12_hive2.1.1_arctic0.5.1.4-release-3.9.16-1.4.11.0.1.jar:?]
	at com.netease.arctic.flink.lookup.BasicLookupFunction.lookup(BasicLookupFunction.java:189) ~[plugin_ne-flink-1.14.0-1.0.13_scala2.12_hive2.1.1_arctic0.5.1.4-release-3.9.16-1.4.11.0.1.jar:?]
	... 78 more
```


### How should we improve?

ThreadLocal provides an easy-to-use API to confine some values to each thread. This is a reasonable way of achieving thread safety in Java.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Subtasks

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/lookup/BinaryRowDataSerializerWrapper.java
+++ b/mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/lookup/BinaryRowDataSerializerWrapper.java
@@ -33,9 +33,9 @@ import java.io.Serializable;
 
 /**
  * This is a wrapper for {@link BinaryRowDataSerializer}. It is used to serialize and deserialize
- * RowData.
+ * RowData. And serialize and deserialize operations are not thread-safe.
  */
-public class BinaryRowDataSerializerWrapper implements Serializable {
+public class BinaryRowDataSerializerWrapper implements Serializable, Cloneable {
 
   private static final long serialVersionUID = 1L;
   protected BinaryRowDataSerializer serializer;
@@ -72,5 +72,10 @@ public class BinaryRowDataSerializerWrapper implements Serializable {
     }
     inputView.setBuffer(recordBytes);
     return serializer.deserialize(inputView);
+  }
+
+  @Override
+  public BinaryRowDataSerializerWrapper clone() {
+    return new BinaryRowDataSerializerWrapper(schema);
   }
 }

--- a/mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/lookup/RocksDBCacheState.java
+++ b/mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/lookup/RocksDBCacheState.java
@@ -66,9 +66,15 @@ public abstract class RocksDBCacheState<V> {
 
   protected final String columnFamilyName;
   protected final ColumnFamilyHandle columnFamilyHandle;
-  protected BinaryRowDataSerializerWrapper keySerializer;
+  protected ThreadLocal<BinaryRowDataSerializerWrapper> keySerializerThreadLocal =
+      new ThreadLocal<>();
 
-  protected BinaryRowDataSerializerWrapper valueSerializer;
+  protected ThreadLocal<BinaryRowDataSerializerWrapper> valueSerializerThreadLocal =
+      new ThreadLocal<>();
+
+  protected final BinaryRowDataSerializerWrapper keySerializer;
+
+  protected final BinaryRowDataSerializerWrapper valueSerializer;
   private ExecutorService writeRocksDBService;
   private final AtomicBoolean initialized = new AtomicBoolean(false);
   private final AtomicBoolean closed = new AtomicBoolean(false);
@@ -127,7 +133,10 @@ public abstract class RocksDBCacheState<V> {
 
   @VisibleForTesting
   public byte[] serializeKey(RowData key) throws IOException {
-    return serializeKey(keySerializer, key);
+    if (keySerializerThreadLocal.get() == null) {
+      keySerializerThreadLocal.set(keySerializer.clone());
+    }
+    return serializeKey(keySerializerThreadLocal.get(), key);
   }
 
   @VisibleForTesting

--- a/mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/lookup/RocksDBRecordState.java
+++ b/mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/lookup/RocksDBRecordState.java
@@ -140,10 +140,17 @@ public class RocksDBRecordState extends RocksDBCacheState<byte[]> {
   }
 
   private byte[] serializeValue(RowData value) throws IOException {
-    return valueSerializer.serialize(value);
+    return valueSerializer().serialize(value);
   }
 
   private RowData deserializeValue(byte[] recordBytes) throws IOException {
-    return valueSerializer.deserialize(recordBytes);
+    return valueSerializer().deserialize(recordBytes);
+  }
+
+  private BinaryRowDataSerializerWrapper valueSerializer() {
+    if (valueSerializerThreadLocal.get() == null) {
+      valueSerializerThreadLocal.set(valueSerializer.clone());
+    }
+    return valueSerializerThreadLocal.get();
   }
 }

--- a/mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/lookup/RocksDBSetSpilledState.java
+++ b/mixed/flink/flink-common/src/main/java/com/netease/arctic/flink/lookup/RocksDBSetSpilledState.java
@@ -42,7 +42,9 @@ import java.util.concurrent.ConcurrentHashMap;
  */
 public class RocksDBSetSpilledState extends RocksDBCacheState<Set<ByteArrayWrapper>> {
   private static final Logger LOG = LoggerFactory.getLogger(RocksDBSetSpilledState.class);
-  protected BinaryRowDataSerializerWrapper joinKeySerializer;
+  protected ThreadLocal<BinaryRowDataSerializerWrapper> joinKeySerializerThreadLocal =
+      new ThreadLocal<>();
+  private final BinaryRowDataSerializerWrapper joinKeySerializer;
   /** Multi-threads would put and delete the joinKeys and Set<ByteArrayWrapper> in the rocksdb. */
   private final Object rocksDBLock = new Object();
 
@@ -76,7 +78,10 @@ public class RocksDBSetSpilledState extends RocksDBCacheState<Set<ByteArrayWrapp
 
   @Override
   public byte[] serializeKey(RowData key) throws IOException {
-    return serializeKey(joinKeySerializer, key);
+    if (joinKeySerializerThreadLocal.get() == null) {
+      joinKeySerializerThreadLocal.set(joinKeySerializer.clone());
+    }
+    return serializeKey(joinKeySerializerThreadLocal.get(), key);
   }
 
   /**

